src/sparse/sparse_matrix.h

```diff
@@ -0,0 +1,131 @@
+/*!
+ *  Copyright (c) 2022 by Contributors
+ * \file sparse/sparse_matrix.h
+ * \brief DGL C++ sparse matrix header
+ */
+#ifndef DGL_SPARSE_SPARSE_MATRIX_H_
+#define DGL_SPARSE_SPARSE_MATRIX_H_
+
+#include <torch/custom_class.h>
+#include <torch/script.h>
+
+#include <memory>
+
+namespace dgl {
+namespace sparse {
+
+/*! \brief SparseFormat enumeration */
+enum SparseFormat { kCOO, kCSR, kCSC };
+
+/*! \brief CSR sparse structure */
+struct CSR {
+  torch::Tensor indptr, indices;
+  // The element order of the sparse format. In the SparseMatrix, we have data
+  // (value_) for each non-zero values. The elements in sparse formats may not
+  // follow the same order as in the SparseMatrix. So we store `e_order` in a
+  // sparse format to indicate its relative element order to the SparseMatrix.
+  // With `e_order`, we can retrieve the correct data for a sparse format, i.e.,
+  // `value_[e_order]`. If `e_order` is not defined, this sparse format follows
+  // the same element order as in the SparseMatrix.
+  torch::optional<torch::Tensor> e_order;
```

Comment:
Is there a way to avoid `e_order`? This is unusual for a generic sparse matrix class. 
Most of the GNN examples will require SpMM which will need CSR aka sorted version. Can't we always keep a sorted version?

Reply:
With `e_order`, we can get the data in sorted order. Using `e_order` is to avoid heavy data shuffling when creating a sparse format because `data` could be a high dimensional tensor.

Reply:
Let's call it `value_indices`.

Reply:
Some of the members in https://github.com/dmlc/dgl/blob/master/include/dgl/aten/csr.h#L38 should be useful. This will also make it easier to use DGL's existing sparse operators.